Merge openyurt into main (#918)
vhive-serverless/vHive: scripts/openyurt-deployer/README.md

**Description**

1. Prepare system environment for all nodes, installing kubeadm, kubectl, dependencies, etc.
2. On master node, init the cluster using `kubeadm init` and in each worker node, join the initialized cluster.
3. On top of the created cluster, init `openyurt` cluster both on master nodes and worker nodes, then expand to all worker nodes from master nodes.
4. (Optional) Deploy Knative (vHive stock-only mode compatible)

## 3. Usage

```bash
./openyurt_deployer clean # clean the openyurt cluster and restore it to initial state 
```

### 3.1 Preparations 
1. Prepare a cluster with at least two nodes.
2. Change the contents in `conf.json` to following format:
```plaintext
{
  "master": "user@master",
  "workers": {
    "cloud": [
      "user@cloud-0"
    ],
    "edge": [
      "user@edge-0"
    ]
  }
}
```

### 3.2 Run Script

#### 3.2.1 To set up Kubernetes and Kubernetes cluster
```bash
go build .
./openyurt_deployer k8s
```
If it gives out error like: 
```
FATA[0001] Failed to connect to: username@host
```
Please execute:
```
eval `ssh-agent -s` && ssh-add ~/.ssh/<your private key>
```
For example:
```
eval `ssh-agent -s` && ssh-add ~/.ssh/id_rsa
```

> And try again

Reviewer: format this as a note so it stands out from the rest of the instructions

#### 3.2.2 To set up Knative 
```bash
./openyurt_deployer knative
```

#### 3.2.2 To set up Open Yurt 
```bash
./openyurt_deployer openyurt
```

## 4. Demo: Create `NodePool` And Deploy service on it
**Referenced from [`OpenYurt`](https://openyurt.io/docs/user-manuals/workload/node-pool-management)*

The demo would deploy a helloworld function to cloud node pool or edge node pool

Deploy the demo:
```

> ./openyurt_deployer demo-c

Reviewer: ```suggestion
./openyurt_deployer demo-deploy-on-cloud
```

```
or:
```

> ./openyurt_deployer demo-e

Reviewer: ```suggestion
./openyurt_deployer demo-deploy-on-edge
```

# Quick set-up `OpenYurt`

> 

Reviewer: add an overview of the Knative-atop-OpenYurt mode. You can give the architecture overview figure (from our tutorial slides) and a pointer to the tutorial slides.

## 1.1 Introduction

> This program extends [`EasyOpenyurt`](https://github.com/flyinghorse0510/easy_openyurt) to automate the set up process of an `OpenYurt` cluster. 

Reviewer: this is irrelevant. the code in vHive doesn't use any of the old code (I hope)

# Quick set-up `OpenYurt`

![Architecture Diagram](/docs/figures/knative_atop_openyurt.png)

Refer to [Serverless Functions atop of Cloud-Edge Deployment](https://docs.google.com/presentation/d/1QPwwvBhE1dZy430x_wDZPgsUYtlFNlteT9TPk3VbRsY/edit?usp=drive_link) tutorial to understand its application.

## 1.1 Introduction

> It support setting up a Kubernetes cluster using kubeadm and then deploy `OpenYurt` and Knative on it. It is compatible with vHive stock-only mode.

Reviewer 2: ```suggestion
OpenYurt extends Kubernetes from a single cluster to a federation of clusters, offers a seamless way to manage distributed nodes and applications efficiently. This guide outlines how to establish a Kubernetes cluster via kubeadm and then deploy `OpenYurt` and Knative on it, which is compatible with vHive stock-only mode.
```

## 1.2 About [`OpenYurt`](https://openyurt.io/docs/#:~:text=Furthermore%2C%20OpenYurt%20enhances%20node%20reliability,node%20heartbeats%20to%20the%20cloud.)

### Key Features

#### 1. Powerful edge autonomy capability

> `OpenYurt` addresses this issue by implementing a per-node proxy (`YurtHub`) along with local storage to cache the state of the cloud apiserver. Consequently, when a node loses its connection, the cached states remain accessible to Kubelet, `KubeProxy`, and any user Pods.

Reviewer 2: ```suggestion
`OpenYurt` is designed to ensure that applications running at the edge of the network remain operational, even in poor connectivity conditions. `OpenYurt` addresses this issue by implementing a per-node proxy (`YurtHub`) along with local storage to cache the state of the cloud apiserver. Consequently, when a node loses its connection, the cached states remain accessible to Kubelet, `KubeProxy`, and any user Pods.
```

#### 2. Cross `NodePool` network communication capability

> In an edge computing Kubernetes cluster, nodes are often distributed across various geographical regions. Consequently, when relying on a native Container Network Interface (CNI) solution, Pods within different `NodePools` may be unable to communicate using Pod IP, Service IP, or Node IP, particularly if each `NodePool` resides within its own isolated LAN. Raven offers a networking solution that enables `cross-NodePool` communication within an `OpenYurt` cluster.

Reviewer 2: ```suggestion
In an edge computing Kubernetes cluster, nodes are often distributed across various geographical regions. The edge nodes fell within same physical region are referred to as `Pools` in OpenYurt. Consequently, when relying on a native Container Network Interface (CNI) solution, Pods within different `NodePools` may be unable to communicate using Pod IP, Service IP, or Node IP, particularly if each `NodePool` resides within its own isolated LAN. OpenYurt introduces a gateway called `Raven`, which offers a networking solution that enables `cross-NodePool` communication within an `OpenYurt` cluster.
```

#### 3. `Multi-NodePool` Management

> In order to manage applications and traffic in multiple node pools conveniently, `YurtAppSet` and `YurtAppDaemon` are introduced for managing workloads in `multi-nodepool`, and service topology capability for routing traffic in node pool.

Reviewer 2: ```suggestion
In order to manage applications and traffic in multiple node pools conveniently, `YurtAppSet` and `YurtAppDaemon` are introduced for managing workloads in `multi-nodepool`, and service topology capability (geographical location, network latency, etc.) for routing traffic in node pool.
```
